### PR TITLE
switch to commonmark for uniformity and more extensions

### LINKED ISSUE #477: Enable raw HTML in markdown for README.md
So I just filed [a bug against Lens](https://github.com/ekmett/lens/issues/633) because of the errant rendering of README.md on the hackage front page for Lens.  I thought it was some kind of simple error in the markup to be fixed quickly.

Turns out, [the behavior is intentional](https://github.com/haskell/hackage-server/pull/356#issuecomment-102956799): hackage-server explicitly avoids rendering raw HTML.

I propose to reconsider that decision, taking into account the fact that stackage is already rendering raw HTML (e.g., see http://www.stackage.org/package/lens).  It works there, and now lens's page just looks worse on hackage.  It also works on github.  Basically hackage is simply behind.

Is there some compelling benefit?  Security issues?  Seems like a package submitter can only mess up their own page with invalid html,  not anybody else's, so why not?

### LINKED ISSUE #565: README/Changelog rendering: Support a larger subset of GitHub-flavored Markdown
As a large part of Haskell packages is hosted on GitHub it would be very convenient if Hackage's Markdown rendering would support a larger subset of Github-flavored Markdown.

These are some features of Github-flavored Markdown that Hackage currently doesn't support:

* Tables (see e.g. [hpack](http://hackage.haskell.org/package/hpack-0.16.0#readme))
* (Sanitized) [Inline HTML](https://github.com/haskell/hackage-server/issues/477)

### LINKED ISSUE #567: Mathjax in readme's
In haddock it is now possible to use Mathjax, can we also enable in Hackage as well? In particular, I would like to be able to use tex-style math in the Readme section of my package(s).

### LINKED ISSUE #650: Support for reference-style links in readme
The `README.md` file here http://hackage.haskell.org/package/hidden-char-0.1.0.1 is rendered correctly (see under "Readme for hidden-char-0.1.0.1").

I recently published an updated version of this package at https://hackage.haskell.org/package/hidden-char-0.1.0.2 and the `README.md` content is garbled. Specifically the hyperlinks at the bottom are output to the web page and not rendered as links.

### LINKED ISSUE #825: img tags in Markdown are not rendered
See e.g. http://hackage.haskell.org/package/hedgehog-0.6.1#readme.

GitHub [shows the image](https://github.com/hedgehogqa/haskell-hedgehog), Hackage shows HTML code. Though it may be a GitHub-only markdown extension...

### PR DESCRIPTION
**Pull request by gbaz**
resolves #565
resolves #825
resolves #477
resolves #650 
resolves #567 

For math we still need to include the following

- [x] `<script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>`

### CODE REVIEW

**Comment by jgm** [comment]
I see this is stalled.  Is it something I can help with?

**Comment by bbarker** [comment]
Seconded, would love to see this - other than the noted conflict are there any concerns?

**Comment by jgm** [comment]
I note that the build failures have to do with HsOpenSSL, so they don't have anything to do with the specific changes in this PR.

**Comment by gbaz** [comment]
There's no major reason for a stall. We (or more likely I) just need to audit the code to figure out the right places to insert the js, then test against a set of packages making use of markdown to make sure there aren't significant regressions, and then it can merge. Note that the hackage redeploy schedule is pretty infrequent, but there have been enough changes we should look towards one.

If someone wants to pull together a good list of "test packages" that would certainly be appreciated.

**Comment by phadej** [comment]
- https://hackage.haskell.org/package/lens
- https://hackage.haskell.org/package/prolens
- https://hackage.haskell.org/package/hpack-0.16.0#readme (newer versions don't bundle README)
- https://hackage.haskell.org/package/hedgehog-0.6.1#readme

These are few packages from issues this PR is supposed to resolve + e.g. `prolens` has quite fancy README

**Comment by phadej** [comment]
The HsOpenSSL issues seems to be also resolved: https://hackage.haskell.org/package/HsOpenSSL-0.11.5/revisions/

**Comment by phadej** [comment]
That said, I don't see any reason for `hackage-server` to support pre GHC-8.6 compilers at this point.

**Comment by andreasabel** [comment]
Maybe this PR could be merged without the math rendering working yet (if that is what is stalling it), there is plenty other issues fixed by the PR.

**Comment by gbaz** [comment]
@andreasabel it was stalled just because I wasn't in good shape to do active hackage development and this needed at least a few eyeball smoke-tests before blindly merging. I think we're in shape now. Hoping to do another redeploy this weekend.

**Comment by gbaz** [comment]
Ok this should now be live.

File: Distribution/Server/Util/Markdown.hs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                     txt)
+  where txt = T.decodeUtf8With T.lenientDecode . BS.toStrict $ md
+
+renderMarkdownRel :: String -> BS.ByteString -> XHtml.Html
+renderMarkdownRel name md =
+     either (const $ XHtml.pre XHtml.<< T.unpack txt) (XHtml.primHtml . T.unpack . sanitizeBalance . TL.toStrict . renderHHtml) $
+         runIdentity (commonmarkWith (mathSpec <> footnoteSpec <> defaultSyntaxSpec <> gfmExtensions)
+                     name
+                     txt)
+  where txt = T.decodeUtf8With T.lenientDecode . BS.toStrict $ md
```
Comment (andreasabel) [inline Distribution/Server/Util/Markdown.hs:118]: When I was a kid, there were these puzzles "Can you see the difference?" with two identically looking pictures next to each other where you had to stare for minutes to see the stroke missing on one side and another stroke missing on the other side.
Hint: the diff between `renderMarkdown` and `renderMarkdownRel` is one letter (at least that is how far I got from staring for minutes).

File: Distribution/Server/Util/Markdown.hs [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  | otherwise = url
+
+
+renderHHtml :: HHtml () -> TL.Text
+renderHHtml (HHtml x) = renderHtml x
+
+renderMarkdown :: String -> BS.ByteString -> XHtml.Html
+renderMarkdown name md =
+     either (const $ XHtml.pre XHtml.<< T.unpack txt) (XHtml.primHtml . T.unpack . sanitizeBalance . TL.toStrict . (renderHtml :: Html () -> TL.Text)) $
+         runIdentity (commonmarkWith (mathSpec <> footnoteSpec <> defaultSyntaxSpec <> gfmExtensions)
```
Comment (andreasabel) [inline Distribution/Server/Util/Markdown.hs:107]: I suppose `<>` is not commutative.  What happens if we place `defaultSyntaxSpec` last?

File: Distribution/Server/Util/Markdown.hs (hunk 1 above)
Comment (gbaz) [reply Distribution/Server/Util/Markdown.hs:118]: Indeed. One uses HHtml and the other Html. The former renders some links with relative paths to the src dir. A refactor and comments would be welcome.

Comment (andreasabel) [reply Distribution/Server/Util/Markdown.hs:118]: See #1003.

**Pull request merged**